TPP.Core/Moderation/Moderator.cs

```diff
@@ -0,0 +1,172 @@
+        private static readonly Duration InitialTimeoutDuration = Duration.FromMinutes(2);
+        // twitch does not allow timeouts beyond 2 weeks
+        private static readonly Duration MaxTimeoutDuration = Duration.FromDays(14) - Duration.FromSeconds(1);
+
+        private readonly int _freeTimeouts;
+        private readonly float _pointsDecayPerSecond;
+        private readonly int _minPoints;
+        private readonly int _pointsForTimeout;
+        private readonly int _pointsForDelete;
+
+        private readonly Dictionary<User, PointStore> _pointsPerUser = new();
+
+        public Moderator(
+            ILogger<Moderator> logger,
+            IExecutor executor,
+            IImmutableList<IModerationRule> rules,
+            IModLogRepo modLogRepo,
+            IClock clock,
+            int freeTimeouts = 2,
+            float pointsDecayPerSecond = 1f,
+            int minPoints = 20,
+            int pointsForTimeout = 300,
+            int pointsForDelete = 200)
+        {
+            _logger = logger;
+            _executor = executor;
+            _modLogRepo = modLogRepo;
+            _clock = clock;
+            _rules = rules;
+            _freeTimeouts = freeTimeouts;
+            _pointsDecayPerSecond = pointsDecayPerSecond;
+            _minPoints = minPoints;
+            _pointsForTimeout = pointsForTimeout;
+            _pointsForDelete = pointsForDelete;
+        }
+
+        private RuleResult ApplyPoints(User user, int points, string reason)
+        {
+            if (points < _minPoints)
+            {
+                _logger.LogDebug($"Ignoring {points} being issued to {user} for reason '{reason}', " +
+                                 $"because the minimum amount of issuable points is {_minPoints}.");
+                return new RuleResult.Nothing();
+            }
+
+            // clean up expired entries, so we don't leak memory
+            List<User> expiredEntries = _pointsPerUser
+                .Where(kvp => kvp.Value.IsEmpty())
+                .Select(kvp => kvp.Key)
+                .ToList();
+            expiredEntries.ForEach(u => _pointsPerUser.Remove(u));
+
+            if (!_pointsPerUser.TryGetValue(user, out PointStore? store))
+            {
+                store = new PointStore(_clock, _pointsDecayPerSecond);
+                _pointsPerUser[user] = store;
+            }
+
+            store.AddPoints(points, reason);
+            int currentPoints = store.GetCurrentPoints();
+            _logger.LogDebug($"Issued {points} points to {user} for reason '{reason}', " +
+                             $"which now has {currentPoints} total.");
+
+            if (currentPoints >= _pointsForTimeout)
+            {
+                IImmutableList<PointStore.Violation> violations = store.GetTopViolations();
+                _pointsPerUser.Remove(user);
+                string topReasons = string.Join(", and ", violations.Select(v => v.Reason));
```

Comment:
I'm not a grammar expert but IIRC you don't use a comma before the and in the list